src/protocol/context/semi_honest.rs

```diff
@@ -1,44 +1,134 @@
+//! Context for protocol executions suitable for semi-honest security model, i.e. secure against
+//! honest-but-curious adversary parties.
+
 use crate::ff::Field;
 use crate::helpers::messaging::{Gateway, Mesh, TotalRecords};
 use crate::helpers::Role;
 use crate::protocol::context::{
     Context, InstrumentedIndexedSharedRandomness, InstrumentedSequentialSharedRandomness,
-    MaliciousContext,
+    MaliciousContext, MaliciousContextBuf,
 };
 use crate::protocol::malicious::MaliciousValidatorAccumulator;
 use crate::protocol::prss::Endpoint as PrssEndpoint;
 use crate::protocol::{Step, Substep};
 use crate::secret_sharing::replicated::semi_honest::AdditiveShare as Replicated;
-use crate::sync::Arc;
 
+use once_cell::sync::OnceCell;
 use std::marker::PhantomData;
 
-/// Context for protocol executions suitable for semi-honest security model, i.e. secure against
-/// honest-but-curious adversary parties.
-#[derive(Clone, Debug)]
-pub struct SemiHonestContext<'a, F: Field> {
-    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
-    /// may operate with raw references and be more efficient
-    pub(super) inner: Arc<ContextInner<'a>>,
+#[derive(Debug)]
+pub struct ContextStore<'a> {
+    pub(super) static_data: ContextStatic<'a>,
+    // Ensure that the struct is invariant over 'a, because that may be required in the future.
+    pub(super) _dummy: PhantomData<OnceCell<&'a ()>>,
```

Comment:
Presumably this also works, if your goal is just to pull in the `'a` lifetime.

```suggestion
    pub(super) _dummy: PhantomData<&'a ()>,
```

Reply:
There is a subtle difference with and without `OnceCell`. Because the `OnceCell` introduces interior mutability, the struct becomes invariant in the lifetime `'a` rather than covariant in the lifetime `'a`. In the covariant case, a function that takes `ContextStore<'a>` will also accept `ContextStore<'b>` for `'b: 'a`. In the invariant case, the function will only accept exactly `ContextStore<'a>` and not any other lifetime.

Minimal example, if you want to play with it: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=386bd282eb9543dd8b991d7e8c7234fb

Reply:
> (You know, it would be really nice if Rust were to build in some way to express common or shared type constraints. We're starting to get enough type constraints on things that compressing those is likely to help, both in terms of making code easier to type out and in making it less error-prone to define new functions and types.)

Agreed. In some situations a trait can be used for this purpose, but that results in some number of `impl MyTraitUsedOnlyAsBoundsAlias for SomeType {}`, which is kind of ugly, especially for library consumers.

Or a macro, but I'm in camp macro-dubious, so I don't particularly like that either.

Reply:
I looked at your covariance/invariance example and I'm a bit baffled.  What about `OnceCell` makes it behave this way when the simple wrapper implementation doesn't?  Is it the use of `UnsafeCell`?

(Count me as in the macro-dubious camp also.  They tend to be opaque; I'm looking for something that is maximally transparent.)

Reply:
Yes, it is the `UnsafeCell` that triggers this. https://doc.rust-lang.org/nomicon/subtyping.html has a useful table.

There is a discussion to be had whether we need this ability to mutate the context. If we don't, then (modulo some cleanup) I already have a branch that achieves the goal of "eliminate Arcs in contexts" with a significantly smaller change than this. However, absent some even larger refactoring of how we handle state associated with communication channels, I felt like some mutability through the context is likely to be necessary.

There's no reason the marker has to be a OnceCell -- but it does have to be `Sync`, which some of the other likely candidates (`Cell` or `&mut`) are not.